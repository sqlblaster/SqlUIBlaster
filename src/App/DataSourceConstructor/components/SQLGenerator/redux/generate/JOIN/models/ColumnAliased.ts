import { Aggregation } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/AggregationsBuilder/components/AggregationItem/model';
import { ColumnBranch } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/common/ColumnSelector/ColumnItem/SelectedColumn.models';
import { Filter } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/FilterBuilder/components/FilterItem/model';
import { Grouping } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/GroupingsBuilder/components/GroupingItem/model';
import { Order } from 'src/App/DataSourceConstructor/components/SQLBuilder/components/OrdersBuilder/components/OrderItem/model';
import { CustomColumn } from '../../../../../SQLBuilder/components/CustomColumnsBuilder/components/CustomColumnItem/model';

export class ColumnBranchAliased extends ColumnBranch {
  public tableAlias: string = '';
  public lastColumn: ColumnBranchAliased;
  public foreignColumn: ColumnBranchAliased;

  constructor(...inits: Partial<ColumnBranch | ColumnBranchAliased>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}

export interface IAliased {
  column: ColumnBranchAliased | undefined;
}

export interface IHasColumn {
  column: ColumnBranch | undefined;
}

export class CustomColumnAliased extends CustomColumn implements IAliased {
  public column: ColumnBranchAliased;

  constructor(...inits: Partial<CustomColumn>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}

export class FilterAliased extends Filter implements IAliased {
  public column: ColumnBranchAliased;

  constructor(...inits: Partial<Filter>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}

export class AggregationAliased extends Aggregation implements IAliased {
  public column: ColumnBranchAliased | undefined;

  constructor(...inits: Partial<Aggregation>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}

export class GroupingAliased extends Grouping implements IAliased {
  public column: ColumnBranchAliased;

  constructor(...inits: Partial<Grouping>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}

export class OrderAliased extends Order implements IAliased {
  public column: ColumnBranchAliased;

  constructor(...inits: Partial<Order>[]) {
    super(...inits);
    Object.assign(this, ...inits);
  }
}
