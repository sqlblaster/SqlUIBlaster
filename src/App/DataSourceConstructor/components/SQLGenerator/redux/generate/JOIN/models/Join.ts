export class Join {
  public primaryTable: JoinTable;
  public foreignTable: JoinTable;

  constructor(...inits: Partial<Join>[]) {
    Object.assign(this, ...inits);
  }
}

export class JoinTable {
  public name: string;
  public alias: string;
  public keyColumn: string = 'ID';

  constructor(...inits: Partial<JoinTable>[]) {
    Object.assign(this, ...inits);
  }
}
