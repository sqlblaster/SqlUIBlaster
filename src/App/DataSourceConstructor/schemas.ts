// TODO: Move everything below to the separate files to the types folder if this will be more than one display size
export class DatabaseSchema {
  public tables: Table[] = [];

  constructor(...inits: Partial<DatabaseSchema>[]) {
    Object.assign(this, ...inits);
  }
}

export class Table {
  public columns: Column[] = [];
  public primaryKeyColumn: string = '';
  public modelName: string;
  private _name: string = '';
  public get name(): Table['_name'] {
    return this._name;
  }
  public set name(value: Table['_name']) {
    this._name = value.toLowerCase();
    this.modelName = value;
  }

  constructor(...inits: Partial<Table>[]) {
    Object.assign(this, ...inits);
    this.findOutPrimaryKeyColumn();
  }

  private findOutPrimaryKeyColumn() {
    const primaryKeyColumn = this.columns.find(
      column => column.type === 'primary'
    );
    primaryKeyColumn && (this.primaryKeyColumn = primaryKeyColumn.name);
  }
}

export class Column {
  public name: string = '';
  public type: 'string' | 'number' | 'Date' | 'boolean' | 'uuid' | 'primary';
  public enums: string[] | undefined;
  public foreignModelName: string | undefined;
  public _foreignTableName: string | undefined;
  public get foreignTableName(): Column['_foreignTableName'] {
    return this._foreignTableName && this._foreignTableName.toLowerCase();
  }
  public set foreignTableName(value: Column['_foreignTableName']) {
    this._foreignTableName = value && value.toLowerCase();
    this.foreignModelName = value;
  }

  constructor(...inits: Partial<Column>[]) {
    Object.assign(this, ...inits);
  }
}
