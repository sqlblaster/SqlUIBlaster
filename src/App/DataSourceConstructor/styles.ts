import { StyleRulesCallback } from '@material-ui/core';

export type QueryBuilderClassKeys = 'root';

export const queryBuilderStyles: StyleRulesCallback<
  QueryBuilderClassKeys
> = (theme) => ({
  root: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'column'
  }
});
