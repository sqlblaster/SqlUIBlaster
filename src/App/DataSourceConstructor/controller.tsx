import { equals } from 'ramda';
import * as React from 'react';
import { connect } from 'react-redux';
import { State } from 'src/store/models/State';
import {
  DataSourceConstructorDispatchProps,
  DataSourceConstructorOwnProps,
  DataSourceConstructorProps,
  DataSourceConstructorStateProps,
  DSCMapDispatchToProps
} from './props';
import { getNewState, preprocessVariables } from './utils';
import { DataSourceConstructor as View } from './view';

export const DataSourceConstructor = connect<
  DataSourceConstructorStateProps,
  DataSourceConstructorDispatchProps,
  DataSourceConstructorOwnProps,
  State
>(state => {
  const { sqlQuery, queryJson, drillDownSqlQuery, drillDownQueryJson } = state;

  return {
    sqlQuery,
    queryJson,
    drillDownSqlQuery,
    drillDownQueryJson,
    rootState: state
  };
}, DSCMapDispatchToProps)(
  class extends React.Component<DataSourceConstructorProps> {
    constructor(props: DataSourceConstructorProps) {
      super(props);

      const {
        databaseSchema,
        initialQuery,
        initialDrillDownQueryJson,
        rootState,
        variables,
        setDatabaseSchema,
        updateRootState
      } = props;

      setDatabaseSchema(databaseSchema);

      initialQuery &&
        updateRootState(
          getNewState(initialQuery, {
            ...rootState,
            ...(initialDrillDownQueryJson
              ? { drillDownQueryJson: initialDrillDownQueryJson }
              : {}),
            tables: databaseSchema.tables,
            variables: preprocessVariables(variables)
          })
        );
    }

    public componentWillReceiveProps({
      databaseSchema: nextDatabaseSchema,
      sqlQuery: nextSqlQuery,
      queryJson: nextQueryJson,
      drillDownSqlQuery: nextDrillDownSqlQuery,
      drillDownQueryJson: nextDrillDownQueryJson,
      initialQuery: nextInitialQuery,
      initialDrillDownQueryJson: nextInitialDrillDownQueryJson,
      variables: nextVariables
    }: DataSourceConstructorProps) {
      const {
        databaseSchema,
        sqlQuery,
        initialQuery,
        drillDownSqlQuery,
        initialDrillDownQueryJson,
        variables,
        rootState,
        onQueryChange,
        setDatabaseSchema,
        updateRootState
      } = this.props;

      if (!equals(databaseSchema, nextDatabaseSchema)) {
        setDatabaseSchema(nextDatabaseSchema);
      }

      if (
        !equals(nextInitialQuery, initialQuery) ||
        !equals(nextVariables, variables) ||
        !equals(nextInitialDrillDownQueryJson, initialDrillDownQueryJson)
      ) {
        updateRootState(
          getNewState(nextInitialQuery, {
            ...rootState,
            variables: preprocessVariables(nextVariables)
          })
        );
      }

      if (sqlQuery !== '' && nextSqlQuery !== sqlQuery && onQueryChange) {
        onQueryChange(
          nextSqlQuery,
          nextQueryJson,
          nextDrillDownSqlQuery,
          nextDrillDownQueryJson
        );
      } else if (drillDownSqlQuery !== nextDrillDownSqlQuery && onQueryChange) {
        onQueryChange(
          nextSqlQuery,
          nextQueryJson,
          nextDrillDownSqlQuery,
          nextDrillDownQueryJson
        );
      }
    }

    public render() {
      return <View showQuery={this.props.showQuery} />;
    }
  }
);
