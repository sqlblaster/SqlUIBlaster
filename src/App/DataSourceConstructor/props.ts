import {
  actionCreators,
  PickActionCreators
} from 'src/store/action-creators';
import {
  PickStates,
  State
} from 'src/store/models/State';
import { Variable } from 'src/schema';
import { ClassesProp } from '../utils/classes-prop';
import { QueryJSON } from './components/SQLGenerator/redux/query-json.model';
import { DatabaseSchema } from './schemas';
import { QueryBuilderClassKeys } from './styles';

export interface DataSourceConstructorOwnProps {
  databaseSchema: DatabaseSchema;
  initialQuery?: QueryJSON;
  initialDrillDownQueryJson?: QueryJSON;
  showQuery?: boolean;
  variables?: Variable[];
  onQueryChange?: (
    query: string,
    queryJson: QueryJSON,
    drillDownSqlQuery: string,
    drillDownQueryJson: QueryJSON
  ) => void;
}

export type DataSourceConstructorStateProps = PickStates<
  'sqlQuery' | 'queryJson' | 'drillDownSqlQuery' | 'drillDownQueryJson'
> & {
  rootState: State;
};

export type DataSourceConstructorDispatchProps = PickActionCreators<
  'setDatabaseSchema' | 'updateRootState'
>;

const {
  setDatabaseSchema,
  updateRootState
}: DataSourceConstructorDispatchProps = actionCreators;
export const DSCMapDispatchToProps: DataSourceConstructorDispatchProps = {
  setDatabaseSchema,
  updateRootState
};

export type DataSourceConstructorProps = DataSourceConstructorOwnProps &
  DataSourceConstructorDispatchProps &
  DataSourceConstructorStateProps;

export type DataSourceConstructorViewProps = Pick<
  DataSourceConstructorOwnProps,
  'showQuery'
> &
  ClassesProp<QueryBuilderClassKeys>;
