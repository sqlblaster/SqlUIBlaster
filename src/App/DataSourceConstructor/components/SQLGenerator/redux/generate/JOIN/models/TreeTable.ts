import { Table } from 'src/App/DataSourceConstructor/schemas';

export class TreeTable {
  private static aliasesCountMap: { [TableName: string]: number } = {};

  public name: string;
  public parentColumnName: string;
  public alias: string = '';
  public primaryKeyColumn: string = '';
  public foreignColumns: string[] = [];
  public foreignTreeTables: TreeTable[] = [];

  constructor(...inits: Partial<TreeTable>[]) {
    Object.assign(this, ...inits);
  }

  public static resetAliasesCountMap() {
    TreeTable.aliasesCountMap = {};
  }

  public setName(name: string) {
    this.name = name;
  }

  public addForeignColumn(column: string) {
    this.foreignColumns.push(column);
  }

  public addForeignTable(table: TreeTable) {
    this.foreignTreeTables.push(table);
  }

  public setAlias() {
    if (!this.alias) {
      const tableName = this.name;

      if (!TreeTable.aliasesCountMap[tableName]) {
        TreeTable.aliasesCountMap[tableName] = 1;
      } else {
        TreeTable.aliasesCountMap[tableName] += 1;
      }

      const aliasNumber = TreeTable.aliasesCountMap[tableName];

      this.alias = `${tableName}${aliasNumber}`;
    }
  }

  public setPrimaryKeyColumn(originalTables: Table[]) {
    const originalTable = originalTables.find(
      table => table.name === this.name
    );
    originalTable && (this.primaryKeyColumn = originalTable.primaryKeyColumn);
  }
}
