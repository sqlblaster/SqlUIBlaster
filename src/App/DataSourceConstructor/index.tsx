import * as React from 'react';
import { Provider } from 'react-redux';
import { Action, Store } from 'redux';
import { getConfiguredStore } from 'src/store';
import { State } from 'src/store/models/State';

import { DataSourceConstructor as Controller } from './controller';
import { DataSourceConstructorOwnProps } from './props';

export class DataSourceConstructor extends React.Component<
  DataSourceConstructorOwnProps
> {
  public store: Store<State, Action> = getConfiguredStore();

  public render() {
    return (
      <Provider store={this.store}>
        <Controller {...this.props} />
      </Provider>
    );
  }
}
