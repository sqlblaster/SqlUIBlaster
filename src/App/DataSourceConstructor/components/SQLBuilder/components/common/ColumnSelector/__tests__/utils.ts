import * as casual from 'casual';
import { Column } from 'src/App/DataSourceConstructor/schemas';
import { ColumnBranch } from '../ColumnItem/SelectedColumn.models';
import {
  filterColumns,
  getNewColumnBranch,
  getRootTableBranch,
  sortInOrderForeignColumnsBelow
} from '../utils';

const tableName = 'Users';

describe('getRootTableBranch', () => {
  it('should return instance of ColumnBranch with specified tableName', () => {
    const columnBranch = getRootTableBranch(tableName, tableName);
    expect(columnBranch).toMatchSnapshot();
  });
});

describe('getNewColumnBranch', () => {
  it('should return new ColumnBranch with defined columnName and type', () => {
    const name = 'username';
    const type: Column['type'] = 'number';

    const columnBranch = getNewColumnBranch(
      new ColumnBranch({
        foreignColumn: new ColumnBranch({
          foreignColumn: new ColumnBranch({
            tableName
          })
        })
      }).normalizeLastColumn(),
      new Column({ name, type })
    );

    expect(columnBranch).toMatchSnapshot();
  });
});

describe('sortInOrderForeignColumnsBelow', () => {
  it('should sort in order foreign columns in the bottom', () => {
    const column1 = new Column({ name: 'ID', type: 'primary' });
    const column2 = new Column({
      name: 'ProductID',
      foreignTableName: 'Products'
    });
    const column3 = new Column({ name: 'name' });
    const column4 = new Column({ name: 'AddedBy', foreignTableName: 'Users' });
    const columns = [column1, column2, column3, column4];

    const sortedColumns = sortInOrderForeignColumnsBelow(columns);

    expect(sortedColumns).toMatchSnapshot();
  });
});

describe('filterColumns', () => {
  const column1 = new Column({ name: 'FirstName', type: 'string' });
  const column2 = new Column({ name: 'Created At', type: 'Date' });
  const column3 = new Column({
    name: 'UserID',
    type: 'number',
    foreignTableName: 'Users'
  });
  const column4 = new Column({ name: 'number of name', type: 'number' });
  const columns: Column[] = [column1, column2, column3, column4];
  const tableBranch = new ColumnBranch({
    tableName
  });

  it(`should filter by filter callback
    should not remove foreign columns`, () => {
    const filteredColumns = filterColumns(
      columns,
      ({ columnName, type }) =>
        (type === 'number' || type === 'string') &&
        columnName.toLocaleLowerCase().endsWith('name'),
      tableBranch
    );

    expect(filteredColumns).toMatchSnapshot();
  });
});
