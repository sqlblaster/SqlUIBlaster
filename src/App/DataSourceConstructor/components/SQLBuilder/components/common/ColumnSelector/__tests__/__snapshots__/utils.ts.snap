// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterColumns should filter by filter callback
    should not remove foreign columns 1`] = `
Array [
  Column {
    "isPrimaryKey": false,
    "name": "FirstName",
    "type": "string",
  },
  Column {
    "_foreignTableName": "users",
    "foreignModelName": "Users",
    "isPrimaryKey": false,
    "name": "UserID",
    "type": "number",
  },
  Column {
    "isPrimaryKey": false,
    "name": "number of name",
    "type": "number",
  },
]
`;

exports[`getNewColumnBranch should return new ColumnBranch with defined columnName and type 1`] = `
Object {
  "_foreignTableName": undefined,
  "_lastColumn": Object {
    "_foreignTableName": undefined,
    "_lastColumn": undefined,
    "columnName": "username",
    "foreignColumn": undefined,
    "foreignModelName": undefined,
    "foreignTableName": undefined,
    "tableName": "Users",
    "type": "number",
  },
  "columnName": "",
  "foreignColumn": Object {
    "_foreignTableName": undefined,
    "_lastColumn": undefined,
    "columnName": "",
    "foreignColumn": Object {
      "_foreignTableName": undefined,
      "_lastColumn": undefined,
      "columnName": "username",
      "foreignColumn": undefined,
      "foreignModelName": undefined,
      "foreignTableName": undefined,
      "tableName": "Users",
      "type": "number",
    },
    "foreignModelName": undefined,
    "foreignTableName": undefined,
    "tableName": "",
    "type": undefined,
  },
  "foreignModelName": undefined,
  "foreignTableName": undefined,
  "tableName": "",
  "type": undefined,
}
`;

exports[`getRootTableBranch should return instance of ColumnBranch with specified tableName 1`] = `
Object {
  "_foreignTableName": undefined,
  "_lastColumn": undefined,
  "columnName": "",
  "foreignColumn": undefined,
  "foreignModelName": undefined,
  "foreignTableName": undefined,
  "tableName": "Users",
  "type": undefined,
}
`;

exports[`sortInOrderForeignColumnsBelow should sort in order foreign columns in the bottom 1`] = `
Array [
  Column {
    "isPrimaryKey": true,
    "name": "ID",
  },
  Column {
    "isPrimaryKey": false,
    "name": "name",
  },
  Column {
    "_foreignTableName": "products",
    "foreignModelName": "Products",
    "isPrimaryKey": false,
    "name": "ProductID",
  },
  Column {
    "_foreignTableName": "users",
    "foreignModelName": "Users",
    "isPrimaryKey": false,
    "name": "AddedBy",
  },
]
`;
